### DISCUSSION TITLE
Deploying fleets of micropython devices but with subtlety different configurations

### DISCUSSION THREAD

**Discussion by proffalken** [micropython/micropython#9406] [Using MicroPython] [answered]
Hey all,

I'm building a solution using esp32 devices and micropython that utilises LoRaWAN.

As part of it, each device needs to be given a distinct name and LoRaWAN configuration (DevEUI, AppSKey etc).

What's the best way to do this in micropython?

I'm currently flashing the code using Ampy, but I'm wondering if there's a better solution than manually updating the config.py file, copying the code over, and restarting the device?

**Comment by peterhinch** [comment]
Unless you can derive things from `machine.unique_id()` or the MAC address I think you're stuck with a `config.py`. There is also [non-volatile storage](http://docs.micropython.org/en/latest/library/esp32.html?#non-volatile-storage) but I'm not sure how you'd use it in this application...

**Reply by proffalken** [reply]
ok, thanks, I'll take a look

**Comment by cwalther** [comment, marked as answer]
Maybe have the device itself connect to a server on the first boot to retrieve its personal _config.py_ (identified by sending along the MAC address or unique ID with the request)? You could put that code in the initial _config.py_ that is then replaced by the real one.

**Reply by proffalken** [reply]
Thanks, but in this case that's not possible. 

The reason I'm using LoRaWAN and not WiFi is because I can't get a WiFi signal to travel 5 miles, but I can do that with LoRaWAN!

It's looking like it's a manual change to the config for each device at the moment.

If I were writing this in C++ with platformio.org then I'd be exporting the variables into the environment for each individual board (easy to do based on a "for" loop) - I guess I could rewrite parts of the config file based on the environment using Jinja2 or similar, but that feels reasonably heavy for this...

**Reply by cwalther** [reply]
I meant while it’s on your desk, before it’s deployed 5 miles into the wild.

**Reply by proffalken** [reply]
> I meant while it’s on your desk, before it’s deployed 5 miles into the wild.

Ah, yes, a very good idea!

So basically a `requests.get()` in boot.py to pull down the config.py file if it doesn't already exist.

That would work...

Thanks!

**Reply by cwalther** [reply]
If you put it in _config.py_ instead of _boot.py_, it’s out of the way (replaced by the real _config.py_) when the device is deployed and doesn’t take up any space and can’t cause any bugs anymore in the field.

**Reply by proffalken** [reply]
Awesome.

Even better, I've just realised that I can probably reuse some of the code (or at least the logic behind it) from something I wrote about 10 years ago to provision physical servers via Puppet and TFTP!

Thanks for the advice!

**Comment by jimmo** [comment]
> I'm currently flashing the code using Ampy, 

@proffalken  Not your actual question, but have you seen `mpremote` ? It's a lot more sophisticated than ampy and has a few really useful features for development (e.g. `mount` ). https://docs.micropython.org/en/latest/reference/mpremote.html

**Reply by proffalken** [reply]
@jimmo it's definitely part of the original question - ampy is... fine... but I'm sure there are better ways of doing this.

I'll checkout mpremote later on today, thanks!
